orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -1,581 +1,765 @@
-}
+/**
+ * @brief Create OrtValue objects from flatbuffer tensors.
+ *
+ * @param flatbuffer_tensors Flatbuffer tensors.
+ * @param name_to_ort_value Name to OrtValue map to be populated.
+ * @return Status of the operation.
+ */
+Status OrtValuesFromFlatbufferTensors(
+    const flatbuffers::Vector<flatbuffers::Offset<onnxruntime::fbs::Tensor>>& flatbuffer_tensors,
+    InlinedHashMap<std::string, OrtValue>& name_to_ort_value) {
+  for (const auto* fbs_tensor : flatbuffer_tensors) {
+    ORT_RETURN_IF_NOT(fbs_tensor, "Encountered a nullptr flatbuffer tensor. Checkpoint file is invalid.");
+
+    std::string tensor_name;
+    OrtValue ort_value;
+    ORT_RETURN_IF_ERROR(OrtValueFromFlatbufferTensor(*fbs_tensor, tensor_name, ort_value));
+    name_to_ort_value.emplace(std::move(tensor_name), std::move(ort_value));
+  }
 
-void WriteTensorProtoToFile(const PathString& file_path,
-                            const std::vector<ONNX_NAMESPACE::TensorProto>& tensor_protos,
-                            std::string caller_context) {
-  auto file_write_status = WithOpenFile(
-      file_path, false,
-      [&tensor_protos](int fd) {
-        google::protobuf::io::FileOutputStream output{fd};
-        ORT_RETURN_IF_ERROR(WriteProtoMessageSequence(tensor_protos, output));
-        return Status::OK();
-      });
-
-  ORT_ENFORCE(file_write_status.IsOK(), caller_context, " write file failed: ", ToUTF8String(file_path));
+  return Status::OK();
 }
 
-void LoadTensorProtoFromFile(const PathString& file_path,
-                             std::vector<ONNX_NAMESPACE::TensorProto>& tensor_protos,
-                             std::string caller_context) {
-  auto file_read_status = WithOpenFile(
-      file_path, true,
-      [&tensor_protos](int fd) {
-        google::protobuf::io::FileInputStream input{fd};
-        ORT_RETURN_IF_ERROR(ReadProtoMessageSequence(tensor_protos, input));
-        return Status::OK();
-      });
-
-  ORT_ENFORCE(file_read_status.IsOK(), caller_context, " load file failed: ", ToUTF8String(file_path));
-}
+namespace save {
 
-template <typename Func>
-void FilterFilesFromDirectory(const PathString& folder_path, Func func) {
-  LoopDir(folder_path, [&func](const PathChar* filename, OrtFileType file_type) -> bool {
-    if (filename[0] == '.' || file_type == OrtFileType::TYPE_DIR) {
-      return true;
-    }
+/**
+ * @brief Save from a checkpoint flatbuffer to file.
+ * @param checkpoint_path Path to save the checkpoint file.
+ * @param builder Flatbuffer builder containing the checkpoint buffer.
+ * @return Status of the operation.
+ *
+ */
+Status ToFile(const PathString& checkpoint_path, flatbuffers::FlatBufferBuilder& builder) {
+  std::ofstream file(checkpoint_path, std::ios::binary);
+  const uint8_t* buf = builder.GetBufferPointer();
+  int size = builder.GetSize();
+  file.write(reinterpret_cast<const char*>(buf), size);
+  ORT_RETURN_IF_NOT(file, "Failed to save checkpoint to file: ", ToUTF8String(checkpoint_path));
```

Comment:
I am curious, is this the common way we check file handle? is it expected to check the file got update after write, or it is to make sure file is created correctly?

Reply:
Yes. As far as I know, this is standard practice to check if any operation performed on the file is successful. This is because the operation returns a stream and the stream can be queried like a boolean to check if the operation was successful.
https://stackoverflow.com/a/16400549